engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/TypeOfNode.java

```diff
@@ -0,0 +1,108 @@
+import org.enso.interpreter.runtime.type.TypesGen;
+
+@BuiltinMethod(
+    type = "Meta",
+    name = "type_of_builtin",
+    description = "Returns the type of a value.")
+public abstract class TypeOfNode extends Node {
+
+  abstract Object execute(@AcceptsError Object value);
+
+  static TypeOfNode build() {
+    return TypeOfNodeGen.create();
+  }
+
+  @Specialization
+  Object doDouble(double value) {
+    Context ctx = Context.get(this);
+    return ctx.getBuiltins().number().getDecimal();
+  }
+
+  @Specialization
+  Object doLong(long value) {
+    Context ctx = Context.get(this);
+    return ctx.getBuiltins().number().getInteger();
+  }
+
+  @Specialization
+  Object doBigInteger(EnsoBigInteger value) {
+    Context ctx = Context.get(this);
+    return ctx.getBuiltins().number().getInteger();
+  }
+
+  @Specialization(guards = {"interop.isDate(value)", "interop.isTime(value)"})
+  Object doDateTime(Object value, @CachedLibrary(limit = "3") InteropLibrary interop) {
+    Context ctx = Context.get(this);
+    return ctx.getBuiltins().dateTime();
+  }
+
+  @Specialization(
+      guards = {"!interop.isDate(value)", "!interop.isTime(value)", "interop.isTimeZone(value)"})
+  Object doTimeZone(Object value, @CachedLibrary(limit = "3") InteropLibrary interop) {
+    Context ctx = Context.get(this);
+    return ctx.getBuiltins().timeZone();
+  }
+
+  @Specialization(guards = {"interop.isDate(value)", "!interop.isTime(value)"})
+  Object doDate(Object value, @CachedLibrary(limit = "3") InteropLibrary interop) {
+    Context ctx = Context.get(this);
+    return ctx.getBuiltins().date();
+  }
+
+  @Specialization(guards = {"!interop.isDate(value)", "interop.isTime(value)"})
+  Object doTime(Object value, @CachedLibrary(limit = "3") InteropLibrary interop) {
+    Context ctx = Context.get(this);
+    return ctx.getBuiltins().timeOfDay();
+  }
+
+  @Specialization(
+      guards = {
+        "interop.hasMetaObject(value)",
+        "!types.hasType(value)",
+        "!interop.isDate(value)",
+        "!interop.isTime(value)",
+        "!interop.isTimeZone(value)"
+      })
+  Object doMetaObject(
+      Object value,
+      @CachedLibrary(limit = "3") InteropLibrary interop,
+      @CachedLibrary(limit = "3") TypesLibrary types) {
+    try {
+      return interop.getMetaObject(value);
+    } catch (UnsupportedMessageException e) {
+      CompilerDirectives.transferToInterpreter();
+      Builtins builtins = Context.get(this).getBuiltins();
+      throw new PanicException(builtins.error().makeCompileError("invalid meta object"), this);
+    }
+  }
+
+  @Specialization(guards = {"types.hasType(value)", "!interop.isNumber(value)"})
+  Object doType(
+      Object value,
+      @CachedLibrary(limit = "3") InteropLibrary interop,
+      @CachedLibrary(limit = "3") TypesLibrary types) {
+    return types.getType(value);
+  }
+
+  @Fallback
+  Object doAny(Object value) {
+    return Context.get(this).getBuiltins().any();
+  }
```

Comment:
I will always take some default top type rather than having a crash that manifests itself to the user. But that is indeed my preference. 

I can't really come up with any example that would hit this case atm.